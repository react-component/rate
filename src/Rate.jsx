import React from 'react';
import ReactDOM from 'react-dom';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { getOffsetLeft } from './util';
import Star from './Star';

function noop() {
}

export default class Rate extends React.Component {
  static propTypes = {
    disabled: PropTypes.bool,
    value: PropTypes.number,
    defaultValue: PropTypes.number,
    count: PropTypes.number,
    allowHalf: PropTypes.bool,
    style: PropTypes.object,
    prefixCls: PropTypes.string,
    onChange: PropTypes.func,
    onHoverChange: PropTypes.func,
    className: PropTypes.string,
    character: PropTypes.node,
  };

  static defaultProps = {
    defaultValue: 0,
    count: 5,
    allowHalf: false,
    style: {},
    prefixCls: 'rc-rate',
    onChange: noop,
    character: 'â˜…',
    onHoverChange: noop,
  };

  constructor(props) {
    super(props);
    let value = props.value;
    if (value === undefined) {
      value = props.defaultValue;
    }

    this.stars = {};

    this.state = {
      value,
    };
  }

  componentWillReceiveProps(nextProps) {
    if ('value' in nextProps) {
      let value = nextProps.value;
      if (value === undefined) {
        value = nextProps.defaultValue;
      }
      this.setState({
        value,
      });
    }
  }

  onHover = (event, index) => {
    const hoverValue = this.getStarValue(index, event.pageX);
    this.setState({
      hoverValue,
    });
    this.props.onHoverChange(hoverValue);
  }

  onMouseLeave = () => {
    this.setState({
      hoverValue: undefined,
    });
    this.props.onHoverChange(undefined);
  }

  onClick = (event, index) => {
    const value = this.getStarValue(index, event.pageX);
    if (!('value' in this.props)) {
      this.setState({
        value,
      });
    }
    this.onMouseLeave();
    this.props.onChange(value);
  }

  getStarDOM(index) {
    return ReactDOM.findDOMNode(this.stars[index]);
  }

  getStarValue(index, x) {
    let value = index + 1;
    if (this.props.allowHalf) {
      const leftEdge = getOffsetLeft(this.getStarDOM(0));
      const width = getOffsetLeft(this.getStarDOM(1)) - leftEdge;
      if ((x - leftEdge - width * index) < width / 2) {
        value -= 0.5;
      }
    }
    return value;
  }

  saveRef = (index) => (node) => {
    this.stars[index] = node;
  }

  render() {
    const { count, allowHalf, style, prefixCls, disabled, className, character } = this.props;
    const { value, hoverValue } = this.state;
    const stars = [];
    const disabledClass = disabled ? `${prefixCls}-disabled` : '';
    for (let index = 0; index < count; index++) {
      stars.push(
        <Star
          ref={this.saveRef(index)}
          index={index}
          disabled={disabled}
          prefixCls={`${prefixCls}-star`}
          allowHalf={allowHalf}
          value={hoverValue === undefined ? value : hoverValue}
          onClick={this.onClick}
          onHover={this.onHover}
          key={index}
          character={character}
        />
      );
    }
    return (
      <ul
        className={classNames(prefixCls, disabledClass, className)}
        style={style}
        onMouseLeave={disabled ? null : this.onMouseLeave}
      >
        {stars}
      </ul>
    );
  }
}
