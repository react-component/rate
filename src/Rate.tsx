import React from 'react';
import findDOMNode from 'rc-util/lib/Dom/findDOMNode';
import classNames from 'classnames';
import KeyCode from 'rc-util/lib/KeyCode';
import { getOffsetLeft } from './util';
import Star from './Star';
import type { StarProps } from './Star';

function noop() {}

export interface RateProps extends Pick<StarProps, "count" | "character" | "characterRender" | "allowHalf" | "disabled"> {
  value?: number;
  defaultValue?: number;
  allowClear?: boolean;
  style?: React.CSSProperties;
  starsSpacing?: React.CSSProperties;
  prefixCls?: string;
  onChange?: (value: number) => void;
  onHoverChange?: (value: number) => void;
  className?: string;
  tabIndex?: number;
  onFocus?: () => void;
  onBlur?: () => void;
  onKeyDown?: React.KeyboardEventHandler<HTMLUListElement>;
  autoFocus?: boolean;
  direction?: string;
}

interface RateState {
  value: number;
  cleanedValue: number;
  hoverValue?: number;
  focused: boolean;
}

class Rate extends React.Component<RateProps, RateState> {
  static defaultProps = {
    defaultValue: 0,
    count: 5,
    allowHalf: false,
    allowClear: true,
    style: {},
    starsStyle: {},
    prefixCls: 'rc-rate',
    onChange: noop,
    character: 'â˜…',
    onHoverChange: noop,
    tabIndex: 0,
    direction: 'ltr',
  };

  stars: Record<string, Star>;

  rate: HTMLUListElement;

  constructor(props: RateProps) {
    super(props);
    let { value } = props;
    if (value === undefined) {
      value = props.defaultValue;
    }

    this.stars = {};

    this.state = {
      value,
      focused: false,
      cleanedValue: null,
    };
  }

  componentDidMount() {
    const { autoFocus, disabled } = this.props;
    if (autoFocus && !disabled) {
      this.focus();
    }
  }

  onHover = (event: React.MouseEvent<HTMLDivElement>, index: number) => {
    const { onHoverChange } = this.props;
    const hoverValue = this.getStarValue(index, event.pageX);
    const { cleanedValue } = this.state;
    if (hoverValue !== cleanedValue) {
      this.setState({
        hoverValue,
        cleanedValue: null,
      });
    }
    onHoverChange(hoverValue);
  };

  onMouseLeave = () => {
    const { onHoverChange } = this.props;
    this.setState({
      hoverValue: undefined,
      cleanedValue: null,
    });
    onHoverChange(undefined);
  };

  onClick = (event: React.MouseEvent | React.KeyboardEvent, index: number) => {
    const { allowClear } = this.props;
    const { value } = this.state;
    const newValue = this.getStarValue(index, (event as React.MouseEvent).pageX);
    let isReset = false;
    if (allowClear) {
      isReset = newValue === value;
    }
    this.onMouseLeave();
    this.changeValue(isReset ? 0 : newValue);
    this.setState({
      cleanedValue: isReset ? newValue : null,
    });
  };

  onFocus = () => {
    const { onFocus } = this.props;
    this.setState({
      focused: true,
    });
    if (onFocus) {
      onFocus();
    }
  };

  onBlur = () => {
    const { onBlur } = this.props;
    this.setState({
      focused: false,
    });
    if (onBlur) {
      onBlur();
    }
  };

  onKeyDown: React.KeyboardEventHandler<HTMLUListElement> = event => {
    const { keyCode } = event;
    const { count, allowHalf, onKeyDown, direction } = this.props;
    const reverse = direction === 'rtl';
    let { value } = this.state;
    if (keyCode === KeyCode.RIGHT && value < count && !reverse) {
      if (allowHalf) {
        value += 0.5;
      } else {
        value += 1;
      }
      this.changeValue(value);
      event.preventDefault();
    } else if (keyCode === KeyCode.LEFT && value > 0 && !reverse) {
      if (allowHalf) {
        value -= 0.5;
      } else {
        value -= 1;
      }
      this.changeValue(value);
      event.preventDefault();
    } else if (keyCode === KeyCode.RIGHT && value > 0 && reverse) {
      if (allowHalf) {
        value -= 0.5;
      } else {
        value -= 1;
      }
      this.changeValue(value);
      event.preventDefault();
    } else if (keyCode === KeyCode.LEFT && value < count && reverse) {
      if (allowHalf) {
        value += 0.5;
      } else {
        value += 1;
      }
      this.changeValue(value);
      event.preventDefault();
    }
    if (onKeyDown) {
      onKeyDown(event);
    }
  };

  static getDerivedStateFromProps(nextProps: RateProps, state: RateState) {
    if ('value' in nextProps && nextProps.value !== undefined) {
      return {
        ...state,
        value: nextProps.value,
      };
    }
    return state;
  }

  getStarDOM(index: number): HTMLElement {
    return findDOMNode(this.stars[index]);
  }

  getStarValue(index: number, x: number) {
    const { allowHalf, direction } = this.props;
    const reverse = direction === 'rtl';
    let value = index + 1;
    if (allowHalf) {
      const starEle = this.getStarDOM(index);
      const leftDis = getOffsetLeft(starEle);
      const width = starEle.clientWidth;
      if (reverse && x - leftDis > width / 2) {
        value -= 0.5;
      } else if (!reverse && x - leftDis < width / 2) {
        value -= 0.5;
      }
    }
    return value;
  }

  saveRef = (index: number) => (node: Star) => {
    this.stars[index] = node;
  };

  saveRate = (node: HTMLUListElement) => {
    this.rate = node;
  };

  focus() {
    const { disabled } = this.props;
    if (!disabled) {
      this.rate.focus();
    }
  }

  blur() {
    const { disabled } = this.props;
    if (!disabled) {
      this.rate.blur();
    }
  }

  changeValue(value: number) {
    const { onChange } = this.props;
    if (!('value' in this.props)) {
      this.setState({
        value,
      });
    }
    onChange(value);
  }

  render() {
    const {
      count,
      allowHalf,
      style,
      prefixCls,
      disabled,
      className,
      character,
      characterRender,
      tabIndex,
      direction,
      starsSpacing,
    } = this.props;
    const { value, hoverValue, focused } = this.state;
    const stars = [];
    const disabledClass = disabled ? `${prefixCls}-disabled` : '';
    for (let index = 0; index < count; index += 1) {
      stars.push(
        <Star
          ref={this.saveRef(index)}
          index={index}
          count={count}
          disabled={disabled}
          prefixCls={`${prefixCls}-star`}
          allowHalf={allowHalf}
          value={hoverValue === undefined ? value : hoverValue}
          onClick={this.onClick}
          onHover={this.onHover}
          key={index}
          spacing={starsSpacing}
          character={character}
          characterRender={characterRender}
          focused={focused}
        />,
      );
    }
    const rateClassName = classNames(prefixCls, disabledClass, className, {
      [`${prefixCls}-rtl`]: direction === 'rtl',
    });
    return (
      <ul
        className={rateClassName}
        style={style}
        onMouseLeave={disabled ? null : this.onMouseLeave}
        tabIndex={disabled ? -1 : tabIndex}
        onFocus={disabled ? null : this.onFocus}
        onBlur={disabled ? null : this.onBlur}
        onKeyDown={disabled ? null : this.onKeyDown}
        ref={this.saveRate}
        role="radiogroup"
      >
        {stars}
      </ul>
    );
  }
}

export default Rate;
