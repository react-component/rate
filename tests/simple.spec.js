import React from 'react';
import { render, mount } from 'enzyme';
import KeyCode from 'rc-util/lib/KeyCode';
import Rate from '../src';

describe('rate', () => {
  describe('full', () => {
    it('render works', () => {
      const wrapper = render(<Rate count={3} value={1} className="custom" />);
      expect(wrapper).toMatchSnapshot();
    });

    it('render works in RTL', () => {
      const wrapper = render(<Rate count={3} value={1} direction="rtl" className="custom" />);
      expect(wrapper).toMatchSnapshot();
    });

    it('render works with character node', () => {
      const wrapper = render(<Rate count={3} value={1} character="1" />);
      expect(wrapper).toMatchSnapshot();
    });

    it('render works with character function', () => {
      const wrapper = render(
        <Rate
          count={3}
          value={1}
          character={({ index }) => {
            return index + 1;
          }}
        />,
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('click works', () => {
      const handleChange = jest.fn();
      const wrapper = mount(<Rate count={3} value={1} onChange={handleChange} />);
      wrapper
        .find('li > div')
        .at(1)
        .simulate('click');
      expect(handleChange).toBeCalledWith(2);
    });

    it('click works in RTL', () => {
      const handleChange = jest.fn();
      const wrapper = mount(<Rate count={3} value={1} direction="rtl" onChange={handleChange} />);
      wrapper
        .find('li > div')
        .at(1)
        .simulate('click');
      expect(handleChange).toBeCalledWith(2);
    });

    it('support mouseMove', () => {
      const wrapper = mount(<Rate count={3} value={1} />);
      wrapper
        .find('li > div')
        .at(1)
        .simulate('mouseMove');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-full'),
      ).toBe(true);
    });

    it('support mouseMove in RTL', () => {
      const wrapper = mount(<Rate count={3} value={1} direction="rtl" />);
      wrapper
        .find('li > div')
        .at(1)
        .simulate('mouseMove');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-full'),
      ).toBe(true);
    });

    it('support focus and blur', () => {
      const wrapper = mount(<Rate count={3} value={2} />);
      wrapper.simulate('focus');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-focused'),
      ).toBe(true);

      wrapper.simulate('blur');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-focused'),
      ).toBe(false);
    });

    it('support focus and blur in RTL', () => {
      const wrapper = mount(<Rate count={3} value={2} direction="rtl" />);
      wrapper.simulate('focus');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-focused'),
      ).toBe(true);

      wrapper.simulate('blur');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-focused'),
      ).toBe(false);
    });

    describe('support keyboard', () => {
      it('left & right', () => {
        const handleChange = jest.fn();
        const wrapper = mount(<Rate count={3} value={1} onChange={handleChange} />);
        wrapper.simulate('keyDown', { keyCode: KeyCode.LEFT });
        expect(handleChange).toBeCalledWith(0);
        handleChange.mockReset();
        wrapper.simulate('keyDown', { keyCode: KeyCode.RIGHT });
        expect(handleChange).toBeCalledWith(2);
      });

      it('enter', () => {
        const handleChange = jest.fn();
        const wrapper = mount(<Rate count={3} value={1} onChange={handleChange} />);
        wrapper
          .find('li > div')
          .at(2)
          .simulate('keyDown', { keyCode: KeyCode.ENTER });
        expect(handleChange).toBeCalledWith(3);
      });
    });

    describe('support keyboard in RTL', () => {
      it('left & right', () => {
        const handleChange = jest.fn();
        const wrapper = mount(<Rate count={3} value={1} direction="rtl" onChange={handleChange} />);
        wrapper.simulate('keyDown', { keyCode: KeyCode.LEFT });
        expect(handleChange).toBeCalledWith(2);
        handleChange.mockReset();
        wrapper.simulate('keyDown', { keyCode: KeyCode.RIGHT });
        expect(handleChange).toBeCalledWith(0);
      });

      it('enter', () => {
        const handleChange = jest.fn();
        const wrapper = mount(<Rate count={3} value={1} direction="rtl" onChange={handleChange} />);
        wrapper
          .find('li > div')
          .at(2)
          .simulate('keyDown', { keyCode: KeyCode.ENTER });
        expect(handleChange).toBeCalledWith(3);
      });
    });
  });

  describe('allowHalf', () => {
    it('render works', () => {
      const wrapper = render(<Rate count={3} value={1.5} allowHalf className="custom" />);
      expect(wrapper).toMatchSnapshot();
    });

    it('render works more than half ', () => {
      const wrapper = render(<Rate count={3} value={1.6} allowHalf className="custom" />);
      expect(wrapper).toMatchSnapshot();
    });

    it('render works in RTL', () => {
      const wrapper = render(
        <Rate count={3} value={1.5} allowHalf direction="rtl" className="custom" />,
      );
      expect(wrapper).toMatchSnapshot();
    });

    it('click works', () => {
      const wrapper = mount(<Rate count={5} value={4.5} allowHalf />);
      wrapper
        .find('li > div')
        .at(2)
        .simulate('click');
      expect(
        wrapper
          .find('li')
          .at(4)
          .hasClass('rc-rate-star-full'),
      ).toBe(false);
    });

    it('support focus and blur', () => {
      const wrapper = mount(<Rate count={3} value={1.5} allowHalf />);
      wrapper.simulate('focus');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-focused'),
      ).toBe(true);

      wrapper.simulate('blur');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-focused'),
      ).toBe(false);
    });

    it('support focus and blur in RTL', () => {
      const wrapper = mount(<Rate count={3} value={1.5} direction="rtl" allowHalf />);
      wrapper.simulate('focus');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-focused'),
      ).toBe(true);

      wrapper.simulate('blur');
      expect(
        wrapper
          .find('li')
          .at(1)
          .hasClass('rc-rate-star-focused'),
      ).toBe(false);
    });

    it('support keyboard', () => {
      const handleChange = jest.fn();
      const wrapper = mount(<Rate count={3} value={1.5} allowHalf onChange={handleChange} />);
      wrapper.simulate('keyDown', { keyCode: KeyCode.LEFT });
      expect(handleChange).toBeCalledWith(1);
      handleChange.mockReset();
      wrapper.simulate('keyDown', { keyCode: KeyCode.RIGHT });
      expect(handleChange).toBeCalledWith(2);
    });

    it('support keyboard in RTL', () => {
      const handleChange = jest.fn();
      const wrapper = mount(
        <Rate count={3} value={1.5} allowHalf direction="rtl" onChange={handleChange} />,
      );
      wrapper.simulate('keyDown', { keyCode: KeyCode.LEFT });
      expect(handleChange).toBeCalledWith(2);
      handleChange.mockReset();
      wrapper.simulate('keyDown', { keyCode: KeyCode.RIGHT });
      expect(handleChange).toBeCalledWith(1);
    });

    it('hover Rate of allowHalf', () => {
      const onHoverChange = jest.fn();
      const wrapper = mount(<Rate count={3} value={1} allowHalf onHoverChange={onHoverChange} />);
      wrapper
        .find('li > div')
        .at(1)
        .simulate('mouseMove', {
          pageX: -1,
        });
      expect(onHoverChange).toHaveBeenCalledWith(1.5);
    });

    it('hover Rate of allowHalf and rtl', () => {
      const onHoverChange = jest.fn();
      const wrapper = mount(
        <Rate count={3} value={1} allowHalf direction="rtl" onHoverChange={onHoverChange} />,
      );
      wrapper
        .find('li > div')
        .at(1)
        .simulate('mouseMove', {
          pageX: 1,
        });
      expect(onHoverChange).toHaveBeenCalledWith(1.5);
    });
  });

  describe('allowClear', () => {
    it('allowClear is false', () => {
      const handleChange = jest.fn();
      const wrapper = mount(
        <Rate count={5} value={1} allowClear={false} onChange={handleChange} />,
      );
      wrapper
        .find('li > div')
        .at(3)
        .simulate('click');
      wrapper
        .find('li > div')
        .at(3)
        .simulate('click');
      expect(handleChange).toBeCalledWith(4);
    });
    it('allowClear is true', () => {
      const handleChange = jest.fn();
      const wrapper = mount(<Rate count={5} value={4} onChange={handleChange} />);
      wrapper
        .find('li > div')
        .at(3)
        .simulate('click');
      expect(handleChange).toBeCalledWith(0);
    });
    it('cleaned star disable hover', () => {
      const wrapper = mount(<Rate count={5} defaultValue={4} />);
      wrapper
        .find('li > div')
        .at(3)
        .simulate('click');
      wrapper
        .find('li > div')
        .at(3)
        .simulate('mouseMove');
      expect(
        wrapper
          .find('li')
          .at(3)
          .hasClass('rc-rate-star-full'),
      ).toBe(false);
    });
    it('cleaned star reset', () => {
      const wrapper = mount(<Rate count={5} defaultValue={4} />);
      wrapper
        .find('li > div')
        .at(3)
        .simulate('click');
      wrapper.find('ul').simulate('mouseLeave');
      wrapper
        .find('li > div')
        .at(3)
        .simulate('mouseMove');
      expect(
        wrapper
          .find('li')
          .at(3)
          .hasClass('rc-rate-star-full'),
      ).toBe(true);
    });
  });

  describe('focus & blur', () => {
    let container;
    beforeEach(() => {
      container = document.createElement('div');
      document.body.appendChild(container);
    });

    afterEach(() => {
      document.body.removeChild(container);
    });

    it('focus()', () => {
      const handleFocus = jest.fn();
      const wrapper = mount(<Rate count={3} value={1} onFocus={handleFocus} />, {
        attachTo: container,
      });
      wrapper.instance().focus();
      expect(handleFocus).toBeCalled();
    });

    it('blur()', () => {
      const handleBlur = jest.fn();
      const wrapper = mount(<Rate count={3} value={1} onBlur={handleBlur} />, {
        attachTo: container,
      });
      wrapper.instance().focus();
      wrapper.instance().blur();
      expect(handleBlur).toBeCalled();
    });

    it('autoFocus', () => {
      const handleFocus = jest.fn();
      mount(<Rate autoFocus count={3} value={1} onFocus={handleFocus} />, { attachTo: container });
      expect(handleFocus).toBeCalled();
    });
  });

  describe('right class', () => {
    it('rtl', () => {
      const wrapper = mount(<Rate count={3} value={1} direction="rtl" />);
      expect(wrapper.find('.rc-rate-rtl').length).toBe(1);
    });
    it('disabled', () => {
      const wrapper = mount(<Rate count={3} value={1} disabled />);
      expect(wrapper.find('.rc-rate-disabled').length).toBe(1);
    });
  });

  describe('events', () => {
    it('onKeyDown', () => {
      const onKeyDown = jest.fn();
      const wrapper = mount(<Rate count={3} onKeyDown={onKeyDown} />);
      wrapper.simulate('keydown');
      expect(onKeyDown).toHaveBeenCalled();
    });
  });
});
