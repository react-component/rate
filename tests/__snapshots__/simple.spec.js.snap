// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rate allowHalf render works 1`] = `
<ul
  class="rc-rate custom"
  role="radiogroup"
  tabindex="0"
>
  <li
    class="rc-rate-star rc-rate-star-full"
  >
    <div
      aria-checked="true"
      aria-posinset="1"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-half rc-rate-star-active"
  >
    <div
      aria-checked="true"
      aria-posinset="2"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="3"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
</ul>
`;

exports[`rate allowHalf render works in RTL 1`] = `
<ul
  class="rc-rate custom rc-rate-rtl"
  role="radiogroup"
  tabindex="0"
>
  <li
    class="rc-rate-star rc-rate-star-full"
  >
    <div
      aria-checked="true"
      aria-posinset="1"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-half rc-rate-star-active"
  >
    <div
      aria-checked="true"
      aria-posinset="2"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="3"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
</ul>
`;

exports[`rate allowHalf render works more than half  1`] = `
<ul
  class="rc-rate custom"
  role="radiogroup"
  tabindex="0"
>
  <li
    class="rc-rate-star rc-rate-star-full"
  >
    <div
      aria-checked="true"
      aria-posinset="1"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-half rc-rate-star-active"
  >
    <div
      aria-checked="true"
      aria-posinset="2"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="3"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
</ul>
`;

exports[`rate full render works 1`] = `
<ul
  class="rc-rate custom"
  role="radiogroup"
  tabindex="0"
>
  <li
    class="rc-rate-star rc-rate-star-full"
  >
    <div
      aria-checked="true"
      aria-posinset="1"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="2"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="3"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
</ul>
`;

exports[`rate full render works in RTL 1`] = `
<ul
  class="rc-rate custom rc-rate-rtl"
  role="radiogroup"
  tabindex="0"
>
  <li
    class="rc-rate-star rc-rate-star-full"
  >
    <div
      aria-checked="true"
      aria-posinset="1"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="2"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="3"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        ★
      </div>
      <div
        class="rc-rate-star-second"
      >
        ★
      </div>
    </div>
  </li>
</ul>
`;

exports[`rate full render works with character function 1`] = `
<ul
  class="rc-rate"
  role="radiogroup"
  tabindex="0"
>
  <li
    class="rc-rate-star rc-rate-star-full"
  >
    <div
      aria-checked="true"
      aria-posinset="1"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        1
      </div>
      <div
        class="rc-rate-star-second"
      >
        1
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="2"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        2
      </div>
      <div
        class="rc-rate-star-second"
      >
        2
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="3"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        3
      </div>
      <div
        class="rc-rate-star-second"
      >
        3
      </div>
    </div>
  </li>
</ul>
`;

exports[`rate full render works with character node 1`] = `
<ul
  class="rc-rate"
  role="radiogroup"
  tabindex="0"
>
  <li
    class="rc-rate-star rc-rate-star-full"
  >
    <div
      aria-checked="true"
      aria-posinset="1"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        1
      </div>
      <div
        class="rc-rate-star-second"
      >
        1
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="2"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        1
      </div>
      <div
        class="rc-rate-star-second"
      >
        1
      </div>
    </div>
  </li>
  <li
    class="rc-rate-star rc-rate-star-zero"
  >
    <div
      aria-checked="false"
      aria-posinset="3"
      aria-setsize="3"
      role="radio"
      tabindex="0"
    >
      <div
        class="rc-rate-star-first"
      >
        1
      </div>
      <div
        class="rc-rate-star-second"
      >
        1
      </div>
    </div>
  </li>
</ul>
`;
